### PR TITLE
Red spell check selection on iOS

### LINKED ISSUE #124404: Spellcheck text selection styling doesn't look right in CupertinoTextField
When you select a misspelled word in a CupertinoTextField, the styling is the same as other text selection (uses the little bars with the circle, and a color supplied by the theme). 

Apr-07-2023 12-25-07

However, on a UIKit/SwiftUI app, if you select a misspelled word, no bars are used and the highlight color is red. 

Hfgbgb

When speaking with developers, text input is something that has been raised as important to "feel native." So, setting the default styling to match native iOS would be great. But, there might be cases where someone wants to customize this styling. Ideally we expose APIs in this widget to control the styling of the selection, and have the default values match those on iOS (maybe this already exists, but I couldn't tell how to do it from a quick glance). 

However, given that it's[ difficult to adap](https://github.com/flutter/flutter/issues/122647)t between CupertinoTextField and TextField, we might consider leveraging TextField for all platforms, but ensure we can add capabilities to match platform-specific styling and behaviors.

cc: @MitchellGoodwin @camsim99

### PR DESCRIPTION
**Pull request by justinmc**
| Native iOS | Flutter before | Flutter now |
| --- | --- | --- |
| image | Screenshot from 2023-04-19 14-52-13 | Screenshot from 2023-04-19 15-07-46|

Fixes https://github.com/flutter/flutter/issues/124404

### CODE REVIEW

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 1]
```diff
@@ -990,7 +1006,7 @@ class SelectionOverlay {
   /// [MagnifierController.shown].
   /// {@endtemplate}
   void showMagnifier(MagnifierInfo initialMagnifierInfo) {
-    if (_toolbar != null || _contextMenuControllerIsShown) {
+    if (_toolbar != null || _contextMenuController.isShown || _spellCheckMenuController.isShown) {
```
Comment (camsim99) [inline packages/flutter/lib/src/widgets/text_selection.dart:1009]: Can you use `toolbarIsVisible` here?

File: packages/flutter/lib/src/widgets/spell_check.dart [hunk 2]
```diff
@@ -30,11 +31,19 @@ class SpellCheckConfiguration {
...
 
   /// The service used to fetch spell check results for text input.
   final SpellCheckService? spellCheckService;
 
+  /// The color the paint the selection highlight when spell check is showing
+  /// suggestions for a misspelled word.
+  ///
+  /// For example, on iOS, the selection appears red while the spell check menu
+  /// is showing.
+  final Color? misspelledSelectionColor;
```
Comment (camsim99) [inline packages/flutter/lib/src/widgets/spell_check.dart:45]: I considered the alternative of not making this configurable, but with the way we create `SpellCheckConfiguration`s in the text fields, I think this makes the most sense :)

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 3]
```diff
@@ -1288,7 +1304,7 @@ class SelectionOverlay {
   // Manages the context menu. Not necessarily visible when non-null.
   final ContextMenuController _contextMenuController = ContextMenuController();
 
-  bool get _contextMenuControllerIsShown => _contextMenuController.isShown;
```
Comment (camsim99) [inline packages/flutter/lib/src/widgets/text_selection.dart:1291]: What were the side effects of using the same `ContextMenuController` for both the text selection toolbar and the spell check suggestions toolbar?

File: packages/flutter/test/cupertino/text_field_test.dart [hunk 4]
```diff
@@ -9289,4 +9289,74 @@ void main() {
     expect(placeholderWidget.overflow, placeholderStyle.overflow);
     expect(placeholderWidget.style!.overflow, placeholderStyle.overflow);
   });
+
+  testWidgets('tapping on a misspelled word on iOS hides the handles and shows red selection', (WidgetTester tester) async {
+    tester.binding.platformDispatcher.nativeSpellCheckServiceDefinedTestValue =
+      true;
+    const Color defaultSelectionColor = Color(0x33007aff);
+    const Color misspelledSelectionColor = Color(0x62ff9699);
```
Comment (camsim99) [inline packages/flutter/test/cupertino/text_field_test.dart:9297]: Nit: consider leaving a comment about where these values came from.

**Review by camsim99 — APPROVED** [review]
Just left nits and questions :)

File: packages/flutter/lib/src/widgets/spell_check.dart (hunk 2 above)
Comment (justinmc) [reply packages/flutter/lib/src/widgets/spell_check.dart:45]: Ok cool, this was one of the main decisions I wanted to get your feedback on.  I think that however we do it, CupertinoTextField needs to tell EditableText what the color should be somehow, so we can't make it completely private.  The other alternative I considered was adding a parameter `spellCheckSelectionColor` to EditableText, but I think this way seems more organized.

Comment (justinmc) [reply packages/flutter/lib/src/widgets/spell_check.dart:45]: Actually before I commit 100% to this, @MitchellGoodwin do you have any thoughts?

So currently TextField and CupertinoTextField read the selection color from DefaultSelectionStyle and then pass it to EditableText as the selectionColor parameter.  This PR adds the ability for CupertinoTextField to change the selection to red for spell check by passing the new misspelledSelectionColor value to EditableText as a part of SpellCheckConfiguration.

The question is, does this align with your vision for theming/styles in Cupertino?

Comment (MitchellGoodwin) [reply packages/flutter/lib/src/widgets/spell_check.dart:45]: Yes, I don't see any issues with that. 👍

File: packages/flutter/test/cupertino/text_field_test.dart (hunk 4 above)
Comment (justinmc) [reply packages/flutter/test/cupertino/text_field_test.dart:9297]: Good idea.

File: packages/flutter/lib/src/widgets/spell_check.dart (hunk 2 above)
Comment (justinmc) [reply packages/flutter/lib/src/widgets/spell_check.dart:45]: Thanks!

File: packages/flutter/lib/src/widgets/text_selection.dart (hunk 1 above)
Comment (justinmc) [reply packages/flutter/lib/src/widgets/text_selection.dart]: Yes, good call!

**Pull request merged**